File: python/src/jobtriage/api/schemas.py
"""Request and response models for the FastAPI HTTP layer."""

import re
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

MAX_TOP_K = 50
MAX_TRIAGE_TOP_K = 10
MAX_DETAILS = 10
MAX_DEADLINE_WINDOW_DAYS = 365
MAX_LOOKUP_TOP_K = 20

ConceptType = Literal['occupation', 'region']

# JobTech taxonomy concept ids are 12-char nanoids, 4-3-3 segments separated by
# underscores (e.g. "X9jv_K2b_m48"). Reject inputs the agent fabricated.
JOBTECH_CONCEPT_ID = re.compile(r'^[A-Za-z0-9]{4}_[A-Za-z0-9]{3}_[A-Za-z0-9]{3}$')


class JobSearchRequest(BaseModel):
    model_config = ConfigDict(extra='forbid')

    occupation_concept_id: str | None = Field(
        default=None,
        description=(
            'JobTech occupation concept id to filter by. Format: '
            '4-3-3 alphanumeric segments separated by underscores '
            '(e.g. "X9jv_K2b_m48").'
        ),
    )
    region: str | None = Field(
        default=None,
        description='JobTech region concept id to filter by.',
    )
    top_k: int = Field(default=10, ge=1, le=MAX_TOP_K)

    @field_validator('occupation_concept_id')
    @classmethod
    def _check_concept_id(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if not JOBTECH_CONCEPT_ID.match(value):
            raise ValueError(
                'occupation_concept_id must be a 12-char JobTech concept id '
                "(e.g. 'X9jv_K2b_m48'). Pass an id from a prior tool result "
                'rather than inventing one.'
            )
        return value


class SemanticSearchRequest(BaseModel):
    model_config = ConfigDict(extra='forbid')

    query: str = Field(min_length=1, max_length=512)
    top_k: int = Field(default=10, ge=1, le=MAX_TOP_K)


class AdSummary(BaseModel):
    model_config = ConfigDict(extra='forbid')

    ad_id: str
    headline: str
    employer_name: str | None = None
    municipality: str | None = None
    application_deadline: str | None = None
    webpage_url: str | None = None


class RankedAd(AdSummary):
    score: float


class AdDetail(AdSummary):
    description_excerpt: str
    occupation_label: str | None = None


class TriagedAd(RankedAd):
    description_excerpt: str


class DeadlineAd(AdSummary):
    days_until_deadline: int


class JobSearchResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    results: list[AdSummary]


class SemanticSearchResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    results: list[RankedAd]


class JobDetailsRequest(BaseModel):
    model_config = ConfigDict(extra='forbid')

    ad_ids: list[str] = Field(min_length=1, max_length=MAX_DETAILS)


class JobDetailsResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    results: list[AdDetail]


class TriageRequest(BaseModel):
    model_config = ConfigDict(extra='forbid')

    query: str = Field(min_length=1, max_length=512)
    top_k: int = Field(default=5, ge=1, le=MAX_TRIAGE_TOP_K)


class TriageResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    results: list[TriagedAd]


class DeadlineRequest(BaseModel):
    model_config = ConfigDict(extra='forbid')

    window_days: int = Field(default=14, ge=1, le=MAX_DEADLINE_WINDOW_DAYS)
    region: str | None = Field(default=None)
    top_k: int = Field(default=10, ge=1, le=MAX_TOP_K)


class DeadlineResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    results: list[DeadlineAd]


class EngagementEntry(BaseModel):
    model_config = ConfigDict(extra='forbid')

    recorded_on: str
    status: str
    note: str = ''


class EngagementStatusResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    ad_id: str
    entries: list[EngagementEntry]


class HealthResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    status: str
    db_path: str
    model_name: str


class LiveAdSummary(AdSummary):
    description_excerpt: str
    occupation_label: str | None = None


class LiveJobSearchRequest(BaseModel):
    model_config = ConfigDict(extra='forbid')

    query: str | None = Field(default=None, max_length=512)
    occupation_concept_id: str | None = Field(default=None)
    region: str | None = Field(default=None)
    top_k: int = Field(default=5, ge=1, le=MAX_TRIAGE_TOP_K)

    @field_validator('occupation_concept_id')
    @classmethod
    def _check_occupation_concept_id(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if not JOBTECH_CONCEPT_ID.match(value):
            raise ValueError(
                'occupation_concept_id must be a 12-char JobTech concept id '
                "(e.g. 'X9jv_K2b_m48'). Resolve via lookupConcept rather than "
                'inventing one.'
            )
        return value


class LiveJobSearchResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    results: list[LiveAdSummary]


class LiveJobDetailsRequest(BaseModel):
    model_config = ConfigDict(extra='forbid')

    ad_ids: list[str] = Field(min_length=1, max_length=MAX_DETAILS)


class LiveJobDetailsResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    results: list[LiveAdSummary]


class Concept(BaseModel):
    model_config = ConfigDict(extra='forbid')

    concept_id: str
    preferred_label: str
    type: ConceptType


class LookupConceptRequest(BaseModel):
    model_config = ConfigDict(extra='forbid')

    query: str = Field(min_length=1, max_length=128)
    top_k: int = Field(default=5, ge=1, le=MAX_LOOKUP_TOP_K)


class LookupConceptResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    results: list[Concept]

File: web/src/lib/api/schemas.ts
import { z } from 'zod'

export const AdSummarySchema = z
  .object({
    ad_id: z.string(),
    headline: z.string(),
    employer_name: z.string().nullable().optional(),
    municipality: z.string().nullable().optional(),
    application_deadline: z.string().nullable().optional(),
    webpage_url: z.string().nullable().optional(),
  })
  .strict()

export const RankedAdSchema = AdSummarySchema.extend({
  score: z.number(),
}).strict()

export const JobSearchRequestSchema = z
  .object({
    occupation_concept_id: z.string().min(1).optional(),
    region: z.string().min(1).optional(),
    top_k: z.number().int().min(1).max(50).default(10),
  })
  .strict()

export const JobSearchResponseSchema = z
  .object({
    results: z.array(AdSummarySchema),
  })
  .strict()

export const SemanticSearchRequestSchema = z
  .object({
    query: z.string().min(1).max(512),
    top_k: z.number().int().min(1).max(50).default(10),
  })
  .strict()

export const SemanticSearchResponseSchema = z
  .object({
    results: z.array(RankedAdSchema),
  })
  .strict()

export const AdDetailSchema = AdSummarySchema.extend({
  description_excerpt: z.string(),
  occupation_label: z.string().nullable().optional(),
}).strict()

export const TriagedAdSchema = RankedAdSchema.extend({
  description_excerpt: z.string(),
}).strict()

export const DeadlineAdSchema = AdSummarySchema.extend({
  days_until_deadline: z.number().int().min(0),
}).strict()

export const JobDetailsRequestSchema = z
  .object({
    ad_ids: z.array(z.string().min(1)).min(1).max(10),
  })
  .strict()

export const JobDetailsResponseSchema = z
  .object({
    results: z.array(AdDetailSchema),
  })
  .strict()

export const TriageRequestSchema = z
  .object({
    query: z.string().min(1).max(512),
    top_k: z.number().int().min(1).max(10).default(5),
  })
  .strict()

export const TriageResponseSchema = z
  .object({
    results: z.array(TriagedAdSchema),
  })
  .strict()

export const DeadlineRequestSchema = z
  .object({
    window_days: z.number().int().min(1).max(365).default(14),
    region: z.string().min(1).optional(),
    top_k: z.number().int().min(1).max(50).default(10),
  })
  .strict()

export const DeadlineResponseSchema = z
  .object({
    results: z.array(DeadlineAdSchema),
  })
  .strict()

export const LiveAdSummarySchema = AdSummarySchema.extend({
  description_excerpt: z.string(),
  occupation_label: z.string().nullable().optional(),
}).strict()

export const LiveJobSearchRequestSchema = z
  .object({
    query: z.string().min(1).max(512).optional(),
    occupation_concept_id: z.string().min(1).optional(),
    region: z.string().min(1).optional(),
    top_k: z.number().int().min(1).max(10).default(5),
  })
  .strict()

export const LiveJobSearchResponseSchema = z
  .object({
    results: z.array(LiveAdSummarySchema),
  })
  .strict()

export const LiveJobDetailsRequestSchema = z
  .object({
    ad_ids: z.array(z.string().min(1)).min(1).max(10),
  })
  .strict()

export const LiveAdDetailsErrorSchema = z
  .object({
    ad_id: z.string().min(1),
    error: z.string().min(1),
  })
  .strict()

export const LiveJobDetailsResponseSchema = z
  .object({
    results: z.array(LiveAdSummarySchema),
    errors: z.array(LiveAdDetailsErrorSchema).default([]),
  })
  .strict()

export const ConceptKindSchema = z.enum(['occupation', 'region'])

export const ConceptSchema = z
  .object({
    concept_id: z.string().min(1),
    preferred_label: z.string().min(1),
    type: ConceptKindSchema,
  })
  .strict()

export const LookupConceptRequestSchema = z
  .object({
    query: z.string().min(1).max(128),
    top_k: z.number().int().min(1).max(20).default(5),
  })
  .strict()

export const LookupConceptResponseSchema = z
  .object({
    results: z.array(ConceptSchema),
  })
  .strict()

export const EngagementEntrySchema = z
  .object({
    recorded_on: z.string(),
    status: z.string(),
    note: z.string(),
  })
  .strict()

export const EngagementStatusRequestSchema = z
  .object({
    ad_id: z.string().min(1),
  })
  .strict()

export const EngagementStatusResponseSchema = z
  .object({
    ad_id: z.string(),
    entries: z.array(EngagementEntrySchema),
  })
  .strict()

export type AdSummary = z.infer<typeof AdSummarySchema>
export type RankedAd = z.infer<typeof RankedAdSchema>
export type AdDetail = z.infer<typeof AdDetailSchema>
export type TriagedAd = z.infer<typeof TriagedAdSchema>
export type DeadlineAd = z.infer<typeof DeadlineAdSchema>
export type JobSearchRequest = z.input<typeof JobSearchRequestSchema>
export type JobSearchResponse = z.infer<typeof JobSearchResponseSchema>
export type SemanticSearchRequest = z.input<typeof SemanticSearchRequestSchema>
export type SemanticSearchResponse = z.infer<
  typeof SemanticSearchResponseSchema
>
export type JobDetailsRequest = z.input<typeof JobDetailsRequestSchema>
export type JobDetailsResponse = z.infer<typeof JobDetailsResponseSchema>
export type TriageRequest = z.input<typeof TriageRequestSchema>
export type TriageResponse = z.infer<typeof TriageResponseSchema>
export type DeadlineRequest = z.input<typeof DeadlineRequestSchema>
export type DeadlineResponse = z.infer<typeof DeadlineResponseSchema>
export type LiveAdSummary = z.infer<typeof LiveAdSummarySchema>
export type LiveJobSearchRequest = z.input<typeof LiveJobSearchRequestSchema>
export type LiveJobSearchResponse = z.infer<typeof LiveJobSearchResponseSchema>
export type LiveJobDetailsRequest = z.input<typeof LiveJobDetailsRequestSchema>
export type LiveAdDetailsError = z.infer<typeof LiveAdDetailsErrorSchema>
export type LiveJobDetailsResponse = z.infer<
  typeof LiveJobDetailsResponseSchema
>
export type Concept = z.infer<typeof ConceptSchema>
export type ConceptKind = z.infer<typeof ConceptKindSchema>
export type LookupConceptRequest = z.input<typeof LookupConceptRequestSchema>
export type LookupConceptResponse = z.infer<typeof LookupConceptResponseSchema>
export type EngagementEntry = z.infer<typeof EngagementEntrySchema>
export type EngagementStatusRequest = z.input<
  typeof EngagementStatusRequestSchema
>
export type EngagementStatusResponse = z.infer<
  typeof EngagementStatusResponseSchema
>
